### ISSUE TITLE
Webcam example plays back very slowly

### ISSUE THREAD

**Issue by benwilber** [jlaine/aiortc#80]
I am playing around with the webcam example and it appears to play back any type of input media in the browser *very* slowly (maybe dropping frames, timecodes are wrong? -- not sure exactly what the problem is).

I've tried several different input media including just a regular mp4 [example mp4 file](https://ll-streams.s3.amazonaws.com/colorbars.mp4).  My ultimate goal is to feed in mpeg-ts/udp and then distribute via webrtc.  I've made some progress on that by modifying the webcam example like:

`player = MediaPlayer("udp://127.0.0.1:9001", format="mpegts")`

and then streaming mpeg-ts via:

`ffmpeg -re -i https://ll-streams.s3.amazonaws.com/colorbars.mp4 -c copy -f mpegts udp://127.0.0.1:9001`

this seems to work but exhibits the slow playback problem.  also webcam.py spams the console with these error messages:

```
non-existing PPS 0 referenced
non-existing PPS 0 referenced
decode_slice_header error
no frame!
non-existing PPS 0 referenced
non-existing PPS 0 referenced
decode_slice_header error
no frame!
```

any information to point me in the right direction would be really appreciated.  this project is awesome btw thanks a lot for it!  I hope to contribute some day

**Comment by jlaine** [comment]
Slow playback is generally a sign that you are unable to encode to VP8 fast enough, I would suggest lowering the resolution.

**Comment by benwilber** [comment]
Hi @jlaine thanks, yes it does appear to be an encoding issue.  Is this doing that encoding in pure python?  I get like 5 fps on my macbook pro.

**Comment by jlaine** [comment]
The encoding is done in a C function call (like), so *should* work well, but I have also witnessed slowdowns, especially on grainy videos. I wonder if the filtering params are right..

**Comment by jlaine** [comment]
Oh wait I only just noticed you are reading from the network. If you're CPUs not maxed out, it might be we are not reading from the network fast enough. Could you try forcing `throttle_playback` to False here:

https://github.com/jlaine/aiortc/blob/c08c547f86570a9f87e5ab53cea9d5ff4aa6f5f3/aiortc/contrib/media.py#L78

**Comment by jlaine** [comment]
For the memory leak issue I suggest you let a process run with a dummy audio / video sources (just use AudioStreamTrack and VideoStreamTrack) and see whether it still OOMs. The vpx codec doesn't use PyAV for encoding, it taps straight into libvpx. The only bit of PyAV you'll use in that simple case is the AudioFrame / VideoFrame classes, which I believe don't have any leak issues. Can you let me know whether that still reproduces the issue?

**Comment by jlaine** [comment]
I ran some additional tests:

- the (unmodified) server example for two hours -> memory usage seems stable

- a minute program reading a .mp4 in a tight loop with PyAV -> I don't see leaks per se, but maybe sprinkling an occasional `gc.collect()` would keep memory usage down

- with the webcam example : if I slow down encoding a bit by introducing some "sleep()" calls, video playback falls behind and memory usage grows indefinitely. This is probably what you are seeing!

Seeing how I coded MediaPlayer, this isn't that surprising as you basically have one thread capturing frames and pushing them into a queue, and another thread which handles the vp8 encoding.. so if vp8 encoding is lagging, the queue grows, and grows..

I'll try:

- to see if there are some settings I can tweak in the vpx encoder to try and speed things up
- implementing some kind of safety mechanism to drop frames if encoding really isn't keeping up

In the meantime if you are willing to sacrifice some video quality, can you try modifying aiortc/codecs/vpx.py and changing:

```self.cfg.rc_target_bitrate = 500```

to something like:

```self.cfg.rc_target_bitrate = 200```
I'll try to see if there are some buttons I can turn

**Comment by jlaine** [comment]
I did some extensive digging into this, as it puzzled me. I played around with VP8 settings every which way, and it seems the encoder does a good job of respecting it's time budget. If we are doing 30fps encoding time has to be below 30ms per frame, and this is what I was seeing for the actual encoding work. However, if I measured the time between when the RTP sender loop submitted a frame for encoding and when the result came back, I got very inconsistent results. Switching back to a simple `loop.run_in_executor` seems to eliminate the problem.
